enclave-tls/src/enclave_quotes/internal/etls_enclave_quote_select.c

```diff
@@ -77,6 +77,15 @@ enclave_tls_err_t etls_verifier_select(etls_core_context_t *ctx,
 {
 	ETLS_DEBUG("selecting the verifier '%s' ...\n", type);
 
+	/* If type is null, it means user doesn't set for verifier type. The type
+	 * will be matched with evidence type automatically later.
+	 */
+	if (!type) {
+		ctx->flags |= ENCLAVE_TLS_CONF_VERPARA_AUTO;
+		ETLS_INFO("the verifier will be matched with evidence later\n");
+		return ENCLAVE_TLS_ERR_NONE;
+	}
```

Comment:
We don't need to do this, because the below logic will chose the highest priority instance, which is a big probability to match the incoming evidence.

Reply:
if user has no setting for verifier parameter, we will select it according evidence typ before verify operation. It will save time

Reply:
Such a scenario is a big probability: on a sgx platform, no selection means sgx-ecdsa-qve is selected automatically, and then receiving a ecdsa evidence hits the selection.

Reply:
The most important is you introduce a new flag ENCLAVE_TLS_CONF_VERPARA_AUTO but it almost does nothing than the current codes.

Reply:
OK I understand what you are trying to resolve. Change to this:
```C
// explicitly specifying verifier is the rare case
if (type)
    ctx->flags |= ENCLAVE_TLS_CONF_VERIFIER_ENFORCED;

// don't break, and thus handle->verifier is not NULL.
```

Reply:
set ctx->flags for ENCLAVE_TLS_CONF_VERPARA_AUTO will be returned directly. It will reduce time of query enclave_quote_ctx and malloc. Also it can solve pointer  leakage.
if type is empty, why not let it check only once at etls_verifier_select in tls_wrapper_verify_certificate_extension?

if (type)
    ctx->flags |= ENCLAVE_TLS_CONF_VERIFIER_ENFORCED; 
It will lead to leakage pointer. For example, if type is empty, then it will allocate a verifier using the highest priority(sgx_ecdsa_qve), but when verify evidence, the actual evidence type is sgx_la, then it will re-chose a new verifier adn query array enclave_quote_ctx again which will be adpat with sgx_la, but the sgx_ecdsa_qve verifier allocated previous will not be freed.

Reply:
Yours:
```C
if (!type) {
  flags |= AUTO;
  return;
}

// save the efforts on the default verifier initialization
```

Mine:
```C
if (type)
  flags |= ENFORCED;

// take the efforts on initialization the default verifier
```

So the difference between us is that you are assuming it is a big probability that the default verifier is pessimistically not the one processing the very first evidence type. Basing on above assumption, thus your implementation can save the efforts against the waste of default verifier initialization taken in the selection phase. However, if the very first evidence type equals to the default verifier type, your implementation has to run a verifier initialization during handling the evidence. In other words, your implementation can save bootstrap time but waste the runtime in big probability.

My point is difference. My portion optimistically assumes the default verifier is the one corresponding to the very first evidence type, so my implementation just labels a flag and continue to initialize the verifier instance. If the very first evidence type matches up the default verifier instance, we are lucky. Occasionally, we will be punished because the  very first evidence type doesn't equal to the default verifier. In other words, my implementation can increase the bootstrap time but save the runtime in big probability.

In rare scenario, your assumption is best, but we need to focus on the big probability situations.